components/empty-state/test/empty-state-text.test.js

```diff
@@ -0,0 +1,61 @@
+import '../empty-state-text.js';
+import '../../button/button-subtle.js';
+import { expect, fixture, html, oneEvent } from '@open-wc/testing';
+
+import { runConstructor } from '../../../tools/constructor-test-helper.js';
+
+describe('d2l-empty-state-text',  () => {
+
+	it('should construct', () => {
+		runConstructor('d2l-empty-state-text');
+	});
+
+	it('should render a span with the given description', async() => {
+		const description = 'There are currently no courses.';
+		const el = await fixture(html`<d2l-empty-state-text description=${description}></d2l-empty-state-text>`);
+		expect(el.shadowRoot.querySelector('span').innerText).to.be.equal(description);
+	});
+
+	it('should render a subtle button with the action-text', async() => {
+		const actionText = 'Create a Course';
+		const el = await fixture(html`<d2l-empty-state-text action-text=${actionText}></d2l-empty-state-text>`);
+		expect(el.shadowRoot.querySelector('d2l-button-subtle')).to.exist;
+		expect(el.shadowRoot.querySelector('d2l-button-subtle').text).to.be.equal(actionText);
+	});
+
+	it('should not render a subtle button if no action-text is given', async() => {
+		const el = await fixture(html`<d2l-empty-state-text description='There are currently no courses.'></d2l-empty-state-text>`);
+		expect(el.shadowRoot.querySelector('d2l-button-subtle')).to.not.exist;
+	});
+
+	it('should render a subtle button from slot if no action-text is given', async() => {
+		const actionText = 'Create a Course';
+		const el = await fixture(html`
+			<d2l-empty-state-text description='There are currently no courses.'>
+				<d2l-button-subtle text=${actionText}></d2l-button-subtle>
+			</d2l-empty-state-text>
+		`);
+		expect(el.querySelector('d2l-button-subtle')).to.exist;
+		expect(el.querySelector('d2l-button-subtle').text).to.be.equal(actionText);
+
+	});
+
+	it('should not render components in slot if action-text is given', async() => {
+		const actionText = 'Create a Course';
+		const el = await fixture(html`
+			<d2l-empty-state-text description='There are currently no courses.' action-text=${actionText}>
+				<d2l-link  href="https://www.d2l.com/">Should Not Exist</d2l-link>
+			</d2l-empty-state-text>
+		`);
+		expect(el.shadowRoot.querySelector('d2l-link')).to.not.exist;
+		expect(el.shadowRoot.querySelector('d2l-button-subtle')).to.exist;
+		expect(el.shadowRoot.querySelector('d2l-button-subtle').text).to.be.equal(actionText);
+	});
```

Comment:
Except for the test for the event firing below, I might ditch the rest of these and just rely on visual-diff to test those cases. They're going to be pretty fragile when the internals of the component change, and aren't really adding much value when visual-diff exists to check the same things.

Reply:
That's a good point, I didn't consider that the visual diffs overlap with many of these. I suppose it is better to just test that the component is working functionally and leave these sorts of tests to the visual diffs. 👍

Reply:
Yeah exactly. Existing examples don't always follow that reasoning (we didn't have visual-diff out of the gate either), but generally yeah if you're testing that setting a certain attribute results in something appearing a certain way or in a certain place then visual-diff is great for that without worrying too much about _how_ that happened DOM & CSS wise.